ProcessMaker/processmaker
# FOUR-8217: Automate Default Template Creations

## Issue & Reproduction Steps
In the current workflow, updating or adding default templates on a server requires manual intervention, significantly impacting engineers' development time. This PR aims to address this issue by introducing an automated process for updating and creating default templates.


## Solution
To achieve this, the following key changes have been implemented:

- Creation of a new artisan command, `php artisan processmaker:sync-default-templates`, which triggers the `SyncDefaultTemplates` job. This job retrieves the index.json data from the [Process Templates Repo](https://github.com/ProcessMaker/process-templates).
- The artisan command also incorporates the `--queue` flag, allowing the job to run on a daily schedule at a random time. This ensures regular synchronization without manual intervention.
- Removal of existing seeded default process templates, which are then replaced with the templates from the [Process Templates Repo](https://github.com/ProcessMaker/process-templates).

# How to Configure:
To successfully configure the automated default template synchronization, follow these steps:

1. Open your `.env` file for your project.
2. Add the following new variables and their respective values:

- `DEFAULT_TEMPLATE_REPO=process-templates`
   Set the value for this variable as the name of the process-templates repository that contains the templates you want to 
   sync with your environment.

- `DEFAULT_TEMPLATE_BRANCH=develop`
    Set the value for this variable as the name of the branch in the process-templates repository that contains the desired 
    templates for your environment.

- ` DEFAULT_TEMPLATE_CATEGORIES=all`
     This variable allows you to specify specific verticals (directories) in the process-templates repository to be synced with 
     your environment. By default, it is set to all, which means all directories in the develop branch will be synced. 
     
     To specify a single directory, set the value to the directory name. For example, to sync the 'banking' directory in the 
      process- templates repository, set `DEFAULT_TEMPLATE_CATEGORIES=banking`

     To sync multiple directories, separate the names with a comma. 
     For example: `DEFAULT_TEMPLATE_CATEGORIES=banking,human-resources,financing`
14. Save the changes to your .env file.
15. Run `php artisan optimize:clear`

ci:package-data-sources:template-sync

ci:deploy




## How to Test

Test Case A: Remove seeded Template and Sync with GitHub

1. Edit an existing default template, such as the New Hire Onboarding template, to become the owner of the template.
4. Run php artisan migrate to trigger the migrations that remove the seeded default templates and synchronize the existing templates in the Process Templates GitHub repo.
5. Verify that the New Hire Onboarding template does not have duplicates and contains the changes you made.

Test Case B: Fresh install 

1. Ensure you have a backup of the data you want to keep.
2. Wipe your database by running php artisan db:wipe --force && php artisan migrate:fresh --seed --force.
3. Run php artisan processmaker:sync-default-templates.
6. Verify that the GitHub repos are synced under the Templates tab.

## Related Tickets & Packages
- Ticket: [FOUR-8217](https://processmaker.atlassian.net/browse/FOUR-8217)
-  [Process Templates Repo](https://github.com/ProcessMaker/process-templates)

## Code Review Checklist



[FOUR-8217]: https://processmaker.atlassian.net/browse/FOUR-8217?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
